ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -837,7 +900,52 @@ int fillFrom(ByteBuffer channelReadBuffer) {
         buf.put(channelReadBuffer);
       }
       if (!buf.hasRemaining()) {
-        onFillComplete(true);
+        if (!passedInBlobProperties.isEncrypted()) {
+          buf.flip();
```

Comment:
Again - not sure why you need to `flip` now if you didn't before.

Reply:
responded above.

Reply:
Flipping at multiple places based on all these conditions is a bit confusing. Any way to simplify this?